internal/provider/provider_framework.go

```diff
@@ -0,0 +1,585 @@
+	}
+
+	return r, err
+}
+
+func resourceDnsImport_framework(id string, client *DNSClient) (dnsConfig, diag.Diagnostics) {
+	var config dnsConfig
+	var diags diag.Diagnostics
+
+	record := id
+	if !dns.IsFqdn(record) {
+		diags.AddError("Error importing DNS record:",
+			fmt.Sprintf("Not a fully-qualified DNS name: %s", record))
+		return config, diags
+	}
+
+	labels := dns.SplitDomainName(record)
+
+	msg := new(dns.Msg)
+
+	var zone *string
+
+Loop:
+	for l := range labels {
+
+		msg.SetQuestion(dns.Fqdn(strings.Join(labels[l:], ".")), dns.TypeSOA)
+
+		r, err := exchange_framework(msg, true, client)
+		if err != nil {
+			diags.AddError("Error querying DNS record:", err.Error())
+			return config, diags
+		}
+
+		switch r.Rcode {
+		case dns.RcodeSuccess:
+
+			if len(r.Answer) == 0 {
+				continue
+			}
+
+			for _, ans := range r.Answer {
+				switch t := ans.(type) {
+				case *dns.SOA:
+					zone = &t.Hdr.Name
+				case *dns.CNAME:
+					continue Loop
+				}
+			}
+
+			break Loop
+		case dns.RcodeNameError:
+			continue
+		default:
+			diags.AddError(fmt.Sprintf("Error querying DNS record: %v", r.Rcode), dns.RcodeToString[r.Rcode])
+			return config, diags
+		}
+	}
+
+	if zone == nil {
+		diags.AddError("Error querying DNS record:",
+			fmt.Sprintf("No SOA record in authority section in response for %s", record))
+		return config, diags
+	}
+
+	common := dns.CompareDomainName(record, *zone)
+	if common == 0 {
+		diags.AddError("Error querying DNS record:",
+			fmt.Sprintf("DNS record %s shares no common labels with zone %s", record, *zone))
+		return config, diags
+	}
+
+	config.Zone = *zone
+
+	if name := strings.Join(labels[:len(labels)-common], "."); name != "" {
+		config.Name = name
+	}
+
+	return config, nil
+}
+
+func resourceFQDN_framework(config dnsConfig) string {
+
+	fqdn := config.Zone
+	if config.Name != "" {
+		fqdn = fmt.Sprintf("%s.%s", config.Name, fqdn)
+	}
+	return fqdn
+}
+
+func resourceDnsRead_framework(config dnsConfig, client *DNSClient, rrType uint16) ([]dns.RR, diag.Diagnostics) {
+	var diags diag.Diagnostics
+	fqdn := resourceFQDN_framework(config)
+
+	msg := new(dns.Msg)
+	msg.SetQuestion(fqdn, rrType)
+
+	r, err := exchange_framework(msg, true, client)
+	if err != nil {
+		diags.AddError("Error querying DNS record:", err.Error())
+		return nil, diags
+	}
+	switch r.Rcode {
+	case dns.RcodeSuccess:
+		// NS records are returned slightly differently
+		if (rrType == dns.TypeNS && len(r.Ns) > 0) || len(r.Answer) > 0 {
+			break
+		}
+		fallthrough
+	case dns.RcodeNameError:
+		return nil, nil
+	default:
+		diags.AddError(fmt.Sprintf("Error querying DNS record: %v", r.Rcode), dns.RcodeToString[r.Rcode])
+		return nil, diags
+	}
+
+	if rrType == dns.TypeNS {
+		return r.Ns, nil
+	}
+	return r.Answer, nil
+}
+
+func resourceDnsDelete_framework(config dnsConfig, client *DNSClient, rrType uint16) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	fqdn := resourceFQDN_framework(config)
+
+	//nolint:forcetypeassert
```

Comment:
Is `nolint:forcetypeassert` required here? Reason I ask is that removing this directive and running `make lint` locally using `golangci-lint v1.52.2` does not raise any errors.